dhis2/dhis2-core
# feat: prevent simultaneous jobs in a cluster [DHIS2-12018]

### Summary
Implements cluster wide coordination of job execution using (redis) caches.
This affects:
* preventing same job from being started more than once at a same time
* sharing of job running and job completed information (tracking)
* cancelling of jobs

In case of a non cluster setup the the coordination still takes place just that it is stored in a local cache and that is has no actual effect that way.

### Coordination of Starting Jobs
The main mechanism is that a background job is running every 30 seconds to update the caches used to coordinate between nodes. These are set to become outdated after 60 seconds so that should they not be refreshed regularly they out-date quite fast and open for other nodes to run a job of the same type.

If this coordination is "transactional" in the cluster depends on the spring `RedisTemplate` => `BoundValueOperations#setIfAbsent` semantics that we also use in leader election.  

### Coordination of Cancelling Jobs
When a job is cancelled en entry is made in a dedicated cache. This state is checked in the background loop that updates the state of running jobs. If a running job should be cancelled the cancel is made effective locally and the cluster wide cancel request is invalidated. As only one node should be able to run a job of the same type this should be safely cancel all running jobs of that type and return state to an idle state for that type. 

Since the checking loop is only running every 30 seconds there can be a delay in the cancellation request of up to 30 seconds depending on if the request hits the node that is running the job or hits another node.

### Sharing of Job Info
The background loop also transfers local job execution state to the cluster wide state. 
For running jobs this simply overrides/sets the info by type as there is only one node actively running that job anyway.
For completed job the run started most recently per type is kept in the cache.

Again the information can have a delay of up to 30 seconds before it becomes visible for remote notes while actual local state stays immediately visible.

 ### Cache API
To coordinate using the `Cache` API two new methods needed to be added:
* `keys()`
* `putIfAbsent`

### Automatic Testing
Existing tests were adopted. As the main implementation of the cluster feature is located in the `DefaultSchedulingManager` which is **not** used in tests this feature needs manual testing. 

### Manual Testing
Below is just one meaningful scenario of many that can be used to test the different aspects of this.

1. Start server and open the scheduler app.
2. create a job configuration for analytics table job. 
3. start the created job manually
4. check the running status both in overview http://localhost:8080/api/scheduling/running and details http://localhost:8080/api/scheduling/running/ANALYTICS_TABLE 
5. when job is done check the completed status both in overview http://localhost:8080/api/completed/running and details http://localhost:8080/api/scheduling/completed/ANALYTICS_TABLE
